### ISSUE TITLE
How to do dynamic object creation

### ISSUE THREAD

**Issue by paulvt** [diwic/dbus-rs#144]
This is rather more of a question than an issue. I have been struggling to create a D-Bus server that can dynamically create extra objects when callbacks from external libraries are triggered.

For example, consider the following object structure:
```
/foo/bar/Manager
/foo/bar/Manager/Item345
/foo/bar/Manager/Item678
...
```

At start, when the connection iteration is started, only the manager exists. At later points I would like to add the object `/foo/bar/Manager/Item345` but it seems that the introspection data is not updated if one just creates an object to the connection as in the examples:

```rust
let f = tree::Factory::new_fn();
let item_tree = f.tree(()).add(
    f.object_path(format!("/foo/bar/Node{}", item_id, ()))
        .introspectable()
        .add(some_iface);
);
item_tree.set_registered(&connection, true).unwrap();
connection.add_handler(item_tree);
```

When introspecting the location `/foo/bar/Manager`, there is no node `Item345` present, so it is not reachable nor visible. This is probably because `/foo/bar/Manager/Item345` was added as a separate tree?

An idea I had was to use the original tree object that was used to build the "Manager" and append to it, but the problem is that this got moved with `connection.add_handler(manager_tree)` and I cannot reach it.

The `adv_server` example (and the others) work from the assumption that all data is known beforehand, however, this is not always the case.  How to add objects at runtime properly? I am willing to add a `dyn_server` example once I get this working.

**Comment by diwic** [comment]
Introspection is generated dynamically by traversing all paths in the tree, so if we get the new path into the tree then that should solve introspection. 

There should be two ways to get to the tree:

 1. Don't call `connection.add_handler(manager_tree)`. It's just a convenience wrapper. Instead you can call `manager_tree.handle` for every incoming message (`connection.incoming()`) and send the replies (if any) using `connection.send`.

 2. You can call `connection.extract_handler` to get the tree back again. It seems like you cannot downcast the `MsgHandler` to a `Tree` though, so maybe this function is not as useful as I intended...but there is a workaround: wrap your tree in an `Arc`, and add an `Arc<Tree>` clone with `connection.add_handler` (since `MsgHandler` is imlemented for `Arc<Tree>`, not just for `Tree`). When you want to change the tree, you call `connection.extract_handler` and drop the result. Now, with only one reference to the tree, you can call `Arc::get_mut` to get a mutable reference, modify the tree, and when you're done with that, again clone the `Arc` and call `connection.add_handler` on the clone.

In either case remember to call `register_object_path` on the new path because `tree.set_registered` is just a one time operation.

 >  I am willing to add a dyn_server example once I get this working.

That sounds awesome, please do :-)

**Comment by paulvt** [comment]
Oops, it did take me longer to get it all working, but it has been running for a year like a charm! So, thanks for the advise.

I need some more, unfortunately. We recently made the jump from 0.5 to 0.8 and most things changed the same except for setting up the connection and trees. I switched to `dbus::ffidisp::Connection` for now in my port, but I'd really like to use the new API.
However, I cannot seem to find how to manipulate the tree now, or retrieve the tree change it and put it back? How to stop receiving after `tree.start_receive(&connection)`? How is dynamic object creation/removal supposed to work with the new API?
I didn't want to reopen the issue for this but I would appreciate some pointers if you can find the time?

**Comment by diwic** [comment]
`tree.start_receive` is a quick-fix for the simplest case. But it is just a few lines of code and it's easy to write the same thing manually:

        connection.start_receive(message::MatchRule::new_method_call(), Box::new(move |msg, c| {
            if let Some(replies) = tree.handle(&msg) {
                for r in replies { let _ = c.send(r); }
            }
            true
        }));

Now, you can put your tree inside a `Rc`, `RefCell`, `Arc` and/or `Mutex` as you see fit in order to be able to modify the tree between calls to `tree.handle`. And to stop receiving, call `connection.stop_receive`.

**Comment by paulvt** [comment]
This seems to work nicely, thanks. I have one remaining issue tough:

Say, before the first `start_receive` call, I have the tree `/`, `/foo/bar/Manager`, and `/foo/bar/Manager/Item0`. Everything is nice. An introspect on `/` mentions node `foo/bar/Manager` and an introspect on `/foo/bar/Manager` mentions node `Item0`. But then, after item 1 is added: an introspect on `/` mentions node `foo/bar/Manager`, but also `foo/Bar/Manager/item1` suddenly (and not item 0?) An introspect on the manager now also mentions both nodes `Item0` and `Item1` which is correct.

This leads to duplicates in my devel tools (such as d-feet). What I present here is a simplified example because I actually have one more nesting level and now the duplicates start to pile up.
It seems that what I add to the tree after the first `start_receive` call gets a mention in `/` too? How can I prevent this? Or do I need to rebuild the tree?

**Comment by diwic** [comment]
> Say, before the first start_receive call, I have the tree /, /foo/bar/Manager, and /foo/bar/Manager/Item0. Everything is nice. An introspect on / mentions node foo/bar/Manager and an introspect on /foo/bar/Manager mentions node Item0. But then, after item 1 is added: an introspect on / mentions node foo/bar/Manager, but also foo/Bar/Manager/item1 suddenly (and not item 0?) An introspect on the manager now also mentions both nodes Item0 and Item1 which is correct.

In the example above, you have mixed `/foo/bar` and `/foo/Bar`, and D-Bus paths are case sensitive. Assuming you have not done this in your real code, this sounds like an issue that should be fixed. It would help me if you could submit a test case that fails; ideally one that could be added as a unit test to dbus-rs, and open a new issue with this test case. Thanks!

**Comment by paulvt** [comment]
Yes, that was a typo. Ok, I will try to come up with a counter example.

**Comment by paulvt** [comment]
I succeeded! That was quite something :smile:.
I parked it here: https://github.com/LeftClickBV/dbus-rs/tree/dynamic-tree-append-introspect-issue

**Comment by diwic** [comment]
@paulvt Your test case made it easy to fix the bug. Also because I pulled your test case we are reasonably sure it won't come back. Thanks a lot!
